### ISSUE TITLE
Is it possible to use trpc and benefit from edge caching at the same time?

### ISSUE THREAD

**Issue by ambiguous48** [icflorescu/trpc-sveltekit#2]
First off, thank you for this package!

So my current stack is Sveltekit, Prisma, Planetscale & Typescript. I've never used trpc before but I'm interested in learning how to use since it seems(?) like a natural fit that can provide better type safety. At the same time, it would be great to take advantage of edge caching from providers such as Vercel or Cloudflare. 

Would that be possible? The reason I ask is because the docs mention that this package requires the use of `@sveltejs/node-adapter`, as opposed to the vercel or cloudflare adapter.

Because I've never used trpc before, the answer to my previous question might be extremely obvious or outright impossible. Even if it requires changing the structure of my app (splitting it into multiple parts, etc), I'd be grateful for any suggestions/advice.

Ultimately, I'm trying to figure out if it's a good idea to go down this rabbithole.

**Comment by icflorescu** [comment]
First off, thank you for your kind words! I'm glad you find it useful (spreading the word would be helpful 🙏 😉)

TL;DR - I think it is, at least on Vercel. Honestly, the only reason I specifically mentioned `@sveltejs/node-adapter` in the docs is that I did get a chance to test in on other environments at the time of writing.

The package **does work on Vercel**, though; I've tested it today and everything seems to be fine, with the same observations regarding Prisma & superjson setup; you cannot use destructuring when importing the Prisma client, and you must pay special attention when packaging for production in `svelte.config.js`. Those considerations are actually related to [Vite](https://vitejs.dev/), not [tRPC](https://trpc.io/), and I guess they do apply for any project, regardless of whether you're using `trpc-sveltekit` or not.

I can't say for sure **in relation to Cloudflare, but I'm guessing it won't work** - afaik their workers aren't exactly Node.js environments, and tRPC does need a Node.js environment, at least at the moment. Things could change, though. See [this issue](https://github.com/trpc/trpc/issues/1375) for more info.

Now, regarding Edge Cashing, some criteria have to be met, see the [Edge Cache for Serverless Functions](https://vercel.com/docs/concepts/functions/edge-caching) page in their docs. In order to satisfy the requirements, you'd have to set some headers in the response - see [Using responseMeta to cache responses](https://trpc.io/docs/caching#using-responsemeta--to-cache-responses) in tRPC docs.

That's why I've exposed the `reponseMeta` function in `trpc-sveltekit`. It's not yet in the docs - I'll have to update those, maybe create a wiki because the README is long enough already - but it's there.

So, you could initialize your handler like this:

```ts
// src/hooks.ts
import { createContext, router } from '$lib/trpcServer';
import { createTRPCHandle } from 'trpc-sveltekit';

export const handle = createTRPCHandle({
  router,
  createContext,
  responseMeta({ type, errors }) {
    if (type === 'query' && errors.length === 0) {
      const ONE_DAY_IN_SECONDS = 60 * 60 * 24;
      return {
        headers: {
          'cache-control': `s-maxage=1, stale-while-revalidate=${ONE_DAY_IN_SECONDS}`
        }
      };
    }
    return {};
  }
});

```

My current stack is React, TypeScript, Prisma and PostgreSQL. But I'm one of those people looking enviously at Svelte... hence this package. So, *going down the rabbithole* is how we progress.

**Comment by icflorescu** [comment]
I'm gonna pin this until I'll be able to update the docs or create a wiki...

**Comment by icflorescu** [comment]
Docs updated, closing.

**Comment by icflorescu** [comment]
You're welcome.

> Community projects such as this or what the folks over at Sveltekit

Mine is an insignifiant contribution, but there's indeed magic in Svelte / Sveltekit... and I want to help whenever I can find the time and energy :-)

**Comment by qudo-code** [comment]
Thanks for this thread! I am going to use this.
